builder/builder.go

```diff
@@ -351,148 +426,390 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+	return nil
 }
 
-// InitMix will initialise a new swupd-client consumable "mix" with the given
-// based Clear Linux version and specified mix version.
-func (b *Builder) InitMix(clearver string, mixver string, all bool, upstreamurl string) error {
-	// Set up metadata
-	err := ioutil.WriteFile(b.Versiondir+"/.clearurl", []byte(upstreamurl), 0644)
+const (
+	// Using the Unicode "Box Drawing" group
+	treeNil = "    "                // "    "
+	treeBar = "\u2502   "           // "│   "
+	treeMid = "\u251C\u2500\u2500 " // "├── "
+	treeEnd = "\u2514\u2500\u2500 " // "└── "
+)
+
+func (b *Builder) buildTreePrintValue(bundle *bundle, level int, levelEnded []bool) string {
+	// Set up the value for this bundle
+	value := bundle.Name
+	if b.isLocalBundle(bundle.Filename) {
+		value += " (local)"
+	} else {
+		value += " (upstream)"
+	}
+
+	if level == 0 {
+		return value
+	}
+
+	var buff bytes.Buffer
+	// Add continuation bars if earlier levels have not ended
+	for i := 0; i < level-1; i++ {
+		if levelEnded[i] {
+			buff.WriteString(treeNil)
+		} else {
+			buff.WriteString(treeBar)
+		}
+	}
+
+	// Add a mid bar or an end bar
+	if levelEnded[level-1] {
+		buff.WriteString(treeEnd)
+	} else {
+		buff.WriteString(treeMid)
+	}
+
+	// Add the actual value
+	buff.WriteString(value)
+
+	return buff.String()
+}
+
+func (b *Builder) bundleTreePrint(set bundleSet, bundle string, level int, levelEnded []bool) {
+
+	fmt.Println(b.buildTreePrintValue(set[bundle], level, levelEnded))
+
+	levelEnded = append(levelEnded, false)
+	last := len(set[bundle].DirectIncludes) - 1
+	for i, include := range set[bundle].DirectIncludes {
+		levelEnded[level] = i == last
+		b.bundleTreePrint(set, include, level+1, levelEnded)
+	}
+}
+
+type listType int
+
+// Enum of available list types
+const (
+	MixList      listType = iota // List bundles in the mix (with includes)
+	LocalList                    // List bundles available locally
+	UpstreamList                 // List bundles available upstream
+)
+
+// ListBundles prints out a bundle list in either a flat list or tree view
+func (b *Builder) ListBundles(listType listType, tree bool) error {
+	var bundles stringSet
+	var err error
+
+	// Get the bundle sets used for processing
+	mixBundles, err := b.getMixBundlesListAsSet()
 	if err != nil {
 		return err
 	}
-	b.Upstreamurl = upstreamurl
-
-	err = ioutil.WriteFile(b.Versiondir+"/.clearversion", []byte(clearver), 0644)
+	localBundles, err := b.getDirBundlesListAsSet(b.LocalBundleDir)
 	if err != nil {
 		return err
 	}
-	b.Clearver = clearver
-
-	err = ioutil.WriteFile(b.Versiondir+"/.mixversion", []byte(mixver), 0644)
+	upstreamBundles, err := b.getDirBundlesListAsSet(getUpstreamBundlesPath(b.UpstreamVer))
 	if err != nil {
 		return err
 	}
-	b.Mixver = mixver
 
-	// Get upstream bundles
-	if err := b.getUpstreamBundles(clearver, true); err != nil {
-		return err
+	// Get the "top level" bundles
+	switch listType {
+	case MixList:
+		bundles = mixBundles
```

Comment:
Move the getDirBundles..() functions into this switch instead of pre-loading them?

Reply:
I rewrote this function probably 7 times, and one of the versions had that. The problem is that I end up needing them around _most_ of the time anyway in order to do the "where does this bundle come from" printing.

- When printing the mix list, you need the `mixBundles` and `localBundles`, but not `upstreamBundles` (though dangit, I just realized I forgot to switch the `b.isLocalBundle(set[bundle].Filename)` call to a simple `if localBundles[set[bundle].Name]` before pushing.
- When printing either the local or upstream list, you need the `mixBundles` and either the `localBundles` and `upstreamBundles`, but not both. So in those cases, the other one is being read in for no reason.

So at one point I had a version that did that juggling, so only the things that needed to be read in were read in, witch each `case` becoming more complicated. I found it a lot harder to follow just looking at it, which is why I pulled it out this way.

I can give it another pass to see if I can simplify it.

Reply:
I just refactored it and will push a new version for you to check out soon.